SDL/SDLJoystick.cpp

```diff
@@ -11,167 +11,209 @@ extern "C" {
 	}
 }
 
-SDLJoystick::SDLJoystick(bool init_SDL ): thread(NULL), running(true) {
+SDLJoystick::SDLJoystick(bool init_SDL ): thread(NULL), running(true), player1Controller(NULL), player1JoyInstanceID(-1) {
 	if (init_SDL)
 	{
-		SDL_setenv("SDL_JOYSTICK_ALLOW_BACKGROUND_EVENTS", "1", 0);
-		SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_VIDEO);
+		SDL_SetHint(SDL_HINT_JOYSTICK_ALLOW_BACKGROUND_EVENTS, "1");
+		SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_VIDEO | SDL_INIT_GAMECONTROLLER);
 	}
+	
+	if (SDL_GameControllerAddMappingsFromFile("assets/gamecontrollerdb.txt") == -1)
+	{
+			printf("Failed to load control pad mappings, please place gamecontrollerdb.txt in your assets directory\n");
+	}
+	else {
+		printf("Loaded game controller mappings for SDL2\n");
+		setUpPlayer1();
+	}
+}
 
+void SDLJoystick::setUpPlayer1()
+{
 	int numjoys = SDL_NumJoysticks();
-	SDL_JoystickEventState(SDL_ENABLE);
+	int joyIndex = -1;
 	for (int i = 0; i < numjoys; i++) {
-		joys.push_back(SDL_JoystickOpen(i));
-//		printf("Initialized joystick %d: %s",i,SDL_JoystickNameForIndex(i));
-		if(strstr(SDL_JoystickNameForIndex(i),"PLAYSTATION(R)3 Controller"))
-			g_Config.bPS3Controller = true;
+		if (joyIndex == -1 && SDL_IsGameController(i))
+		{
+			joyIndex = i;
+			break;
+		}
+	}
+	if (joyIndex != -1)
```

Comment:
The old handling only had mappings for ~4 control pads. Also, it applied the same mapping to all control pads connected. This of course could produced odd behaviour if different types of control pads were hot plugged as the mappings were not re-calculated. The SDL2 controller database supports over 50 control pads for Linux, over 25 for Windows and over 10 for OSX.

It would be possible to assign a default mapping to unknown control pad, but this would be guess work as we would know I idea of how to map buttons and axis etc.